### PR TITLE
constant interpreter: support evaluating generic functions

### PR DESCRIPTION
**Pull request by marcrasi**
This adds support for evaluating generic functions to the constant interpreter.

This makes SILModule::getSILLoader() and getWitnessMethodSubstitutions public so that the interpreter can access them. Maybe we should do something more proper before merging this. Suggestions about what exactly to do would be very helpful!

This is a part of the #19579 mega-patch. See there if you're interested in what comes next.

### CODE REVIEW

File: lib/SILOptimizer/Utils/Devirtualize.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Devirtualize.cpp]: Instead, why not make a public function which devirtualizes an apply of a witness_method, and then the constant evaluator only has to worry about calls of static function references?

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: Typo: "witness"

File: lib/SILOptimizer/Utils/Devirtualize.cpp
Comment (slavapestov) [reply lib/SILOptimizer/Utils/Devirtualize.cpp]: Mandatory inlining has something similar as well, so at this point it looks like this is adding a fourth copy of a very similar piece of code.

**Comment by marcrasi** [comment]
@slavapestov thanks for the suggestion! I will try it out in the tensorflow branch, and then propagate it to this PR. Hopefully in a few days.

Also fyi, I made a nontrivial change to `getSingleWriterAddressValue` in the tensorflow branch in https://github.com/apple/swift/pull/19705. I will propagate that change into this PR at the same time that I propagate @slavapestov's suggestion.

**Comment by ravikandhadai** [comment]
>  I made a nontrivial change to getSingleWriterAddressValue in the tensorflow branch in #19705. I will propagate that change into this PR

@marcrasi Sounds great. Thanks.

File: lib/SILOptimizer/Utils/Devirtualize.cpp
Comment (marcrasi) [reply lib/SILOptimizer/Utils/Devirtualize.cpp]: I set out to do this and discovered an obstacle: I can't actually mutate the SIL to have static function references (like the other APIs in Devirtualize.h do) because the function containing the apply can still be totally generic. The interpreter knows all the concrete types because it arrived at the apply by following a specific path.

I could create a Devirtualize.h public function that takes an ApplyInst and the current substitutions and returns a devirtualized ApplyInst without replacing the existing one. That might be confusingly different from the other Devirtualize.h APIs, and it also wastefully allocates an ApplyInst that is used for a very temporary purpose.

Comment (marcrasi) [reply lib/SILOptimizer/Utils/Devirtualize.cpp]: What's your suggestion?

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: Just using apply->getSubstitutionMap() should be sufficient -- you don't have to build a new one since it should have the same generic signature.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: I guess I don't really understand why this PR is adding so much new code (lookupOrLinkWitnessTable() and getWitnessMethodSubstitutions()) when very similar logic already exists in the devirtualizer.

You bring up a very good point that while the devirtualizer rewrites instructions in place, the evaluator looks at but does not mutate the AST. However it should be possible to come up with some common pieces that can be reused -- for example, a common utility returning the SILFunction and SubstitutionMap for a devirtualized call. The optimizer can build an ApplyInst from that, and the evaluator can proceed to evaluate the body of the SILFunction.

I'm generally pessimistic about PRs that add TODO/FIXME comments because in practice we rarely see them get addressed, as everyone is keen to move on to the next feature...

Comment (slavapestov) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Sounds good!

Comment (ravikandhadai) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: I have some suggestions for addressing @slavapestov 's concerns.   It seems to me that `lookupOrLinkWitnessTable()` functionality is partly available in the `SILLinkerVisitor::visitProtocolConformance`. Apparently, this is the only other place where we _create_ a witness table declaration for a protocol conformance in case we do not find a witness table for the conformance. (Even though somewhat similar logic exist in other places, it seems to me that there is an assumption that the witness table declaration must exist for a protocol in other places.)  But, much of what visitProtocolConformance does, such as deserializing any witness methods and  visiting related protocol conformances, are not needed for the evaluator. (Deserialization of the witness method would happen later during evaluation, only if the function is actually invoked during evaluation.)  Therefore, it seems to me that  we could factor out this common functionality, shown below, from `visitProtocolConformance`.

```swift
SILWitnessTable *lookupOrCreateWitnessTable(ProtocolConformance *C, bool mustDeserialize,           
    SILModule &Mod) {
  SILWitnessTable *WT = Mod.lookUpWitnessTable(C, true);

  // If we don't find any witness table for the conformance,
  // try to create a witness table declaration and try again.
  if (!WT) {
    Mod.createWitnessTableDeclaration(C,
        getLinkageForProtocolConformance(
                          C->getRootNormalConformance(), NotForDefinition));

    // Adding the declaration may allow us to now deserialize the body.
    // Force the body if we must deserialize this witness table.
    if (mustDeserialize) {
      WT = Mod.lookUpWitnessTable(C, true);
      assert(WT && WT->isDefinition()
             && "unable to deserialize witness table when we must?!");
    } else {
      return WT;
    }
  }
  return WT;
}
```
I am not sure where this function should go (Perhaps in the SILLinker.cpp itself or SILModule.cpp?).  If we have this factored out, both the evaluator and `SILLinkerVisitor::visitProtocolConformance` can use it.

File: lib/SILOptimizer/Utils/Devirtualize.cpp
Comment (slavapestov) [reply lib/SILOptimizer/Utils/Devirtualize.cpp]: @ravikandhadai @marcrasi I think refactoring the devirtualizer to return a `SILFunction *` together with a SubstitutionMap might be the way to go. You're right that allocating a new ApplyInst might be wasteful if you're not actually mutating the SIL.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (ravikandhadai) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: @slavapestov  Thanks for pointing this out. This copy can be removed.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: You probably want callSubMap = callSubMap.subst(substitutionMap)

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: This will just be the protocol's Self type. You want the concrete Self type?

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: I think you want AI->getSubstitutionMap().lookupConformance()

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: If you have a class method, you also need to do some remapping for vtable substitutions. Eg if I have `class G<T, U> : C<(T, U), Int>`, and I'm calling a method on an instance of G<String, Float>, but its actually implemented on C, I have to remap the substitution map to `C<(String, Float), Int>`.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (slavapestov) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Check out https://github.com/apple/swift/pull/21476 -- I updated lookUpWitnessTable() to create the declaration if one doesn't exist, so you can simplify lookupOrLinkWitnessTable() or perhaps remove it entirely.

Comment (ravikandhadai) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Thanks a lot @slavapestov. This is awesome.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (ravikandhadai) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: @slavapestov Thanks for pointing this out. You are right that this code is quite broken. @marcrasi I could make a test case (shown below) that crashes it. 
 I think we should use: 
`apply->getSubstitutionMap().subst(substitutionMap).lookupConformance(...)` 
This is because `apply->getSubstitutionMap()` would create a map in which `Self` type of the protocol will be bound to the receiver type of the call site. And, composing it with `substitutionMap` will map it to the actual concrete receiver type tracked by the interpreter. 
Therefore, we would get the right concrete conformance. This seems to fix.

Just for the reference, the test case that bring out this problem is shown below:

```swift
protocol Proto {
  func amethod<U>(_ u: U) -> Int
}

struct S : Proto {
  func amethod<U>(_ u: U) -> Int {
    return 0
  }
}

func callMethod<U, T: Proto>(_ a: T, _ u: U) -> Int {
  return a.amethod(u)
}

func testProtocolMethod() {
  let s = S()
  #assert(callMethod(s, 10) == 0)
}
```

**Comment by ravikandhadai** [comment]
@marcrasi I have pushed one commit, so that it would easier to merge this PR once you can take look at this. The new commit updated to work with @slavapestov's changes, fixes some bugs, adds more test cases, and improves the comments a bit.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (ravikandhadai) [inline lib/SILOptimizer/Utils/ConstExpr.cpp]: @marcrasi  I have a suggestion here. It seems `!tupleEltAddrValue.isConstant()` is used to imply that tupleEltAddrValue has no writers at all is only read. (Like the %191 example presented above.) But, we are sort of overloading the meaning of !isConstnat() here kind of mean read only. I wonder whether this could be make explicit by returning e.g. a boolean flag from the function `getSingleWriterAddressValue()` to indicate that no writer was found. This code can them check for this flag.

**Comment by ravikandhadai** [comment]
@marcrasi  I have left a couple of minor comments I noticed while making the changes. Let me know if they make sense. Except that,  this PR  looks good to me.

**Comment by marcrasi** [comment]
Sorry for the delay! I'm starting to look now.

**Comment by ravikandhadai** [comment]
Cool. Thanks for working on this. Feel free to ping me if I could clarify any comments here or help with something.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Should we assert here that the function is not a class method. Is there a way to do that?

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: switched!

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (marcrasi) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: I was able to write a SIL test that exposes a bug that this causes! Will change it and add the test.

File: lib/SILOptimizer/Utils/ConstExpr.cpp
Comment (ravikandhadai) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Looks like this may work: 
`assert(!calleeFnType->getSelfInstanceType()->getClassOrBoundGenericClass(), "class methods are not supported")`

Comment (ravikandhadai) [reply lib/SILOptimizer/Utils/ConstExpr.cpp]: Perhaps this is better:
`assert(!calleeFnType->hasSelfParam() || !calleeFnType->getSelfInstanceType()->getClassOrBoundGenericClass())`

**Comment by marcrasi** [comment]
Thanks for all the help with this one, @slavapestov and @ravikandhadai :)

It looks good to me now. I made a few changes:
* asserted that the method is not a class method
* added SIL tests exercising all the getSingleWriterAddressValue behavior
* made the getSingleWriterAddressValue change suggested by @ravikandhadai, which fixed a bug. (and I added a test exposing this bug)

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test

**Comment by ravikandhadai** [comment]
@marcrasi Thanks for all your help and effort.  I think, once this gets merged, what remains is Strings, Arrays, Enums and Floats. Would it be possible for you to create PRs for these? I foresee that String and Arrays might involve more reviews than the other two, and they are introducing support for the well-known functions. Perhaps it is better to start that off, whenever you can.

**Comment by marcrasi** [comment]
Yes, that's what remains. I will try to start PRs for some of them right now.

**Comment by marcrasi** [comment]
I just updated this PR with the changes from https://github.com/apple/swift/pull/21628.

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Review by devincoughlin — APPROVED** [review]
LGTM. Thank you for adding more tests and documentation!

**Comment by ravikandhadai** [comment]
@swift-ci Please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci Please test and merge

**Pull request merged**
